api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -473,7 +494,11 @@ public NestedField asRequired() {
       if (!isOptional) {
         return this;
       }
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, doc, initialDefault, writeDefault);
+    }
+
+    public NestedField updateWriteDefault(Object newWriteDefault) {
```

Comment:
There's precedent for returning a slightly modified version of the field with the `asRequired` and `asOptional` methods, but this method name doesn't make it clear to me that the original field isn't being modified. I think `withWriteDefault` is slightly better, although not great.